### ISSUE TITLE
🚀 Feature: [kubernetes-* plugins] let authentication use from scratch OIDC providers 

### ISSUE THREAD

**Issue by markussiebert** [backstage/backstage#19836]
### 🔖 Feature description

I want to use "from scratch" oidc provider to authenticate with the kubernetes plugin - thats not possible,
(or it's possible and we have no idea how - and we could improve the documentation on this)

### 🎤 Context

We want to use AWS Cognito as OIDC Provider with our EKS cluster together with backstage and the Kubernetes plugin.

We followed the setup process and added AWS Cognito as a generic OIDC provider to backstage.
Then we tried to add it to the Kubernetes plugin, but ended up with this error:

```KubernetesAuthProviders has no oidcProvider configured for oidc.cognito```

after digging around in the code, we found out, that this error is thrown [here](https://github.com/backstage/backstage/blob/3feb20f7536ac25d8432c5562dc82baa23f0cea0/plugins/kubernetes/src/kubernetes-auth-provider/KubernetesAuthProviders.ts#L93C9-L93C85)

[The constructor/factory](https://github.com/backstage/backstage/blob/3feb20f7536ac25d8432c5562dc82baa23f0cea0/plugins/kubernetes/src/plugin.ts#L82) then disclosed the secret - there is a predefined list that limits the OIDC providers to:

```
gitlab: gitlabAuthApi,
google: googleAuthApi,
microsoft: microsoftAuthApi,
okta: oktaAuthApi,
onelogin: oneloginAuthApi,
```

The [docs](https://backstage.io/docs/features/kubernetes/authentication#client-side-providers) don't give a hint how own providers can be used. We thought we can use every OIDC provider - even those from scratch.
[other docs](https://backstage.io/docs/features/kubernetes/configuration/#servicelocatormethod)

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by jamieklassen** [comment]
As you noticed, the Kubernetes plugin defines a default API factory for the `kubernetesAuthProvidersApiRef` here:

https://github.com/backstage/backstage/blob/d2af76ffb6a105dfec1310733f5ca8e316872116/plugins/kubernetes/src/plugin.ts#L66-L96

but as an integrator you can override it with your own implementation by customizing the `apis` parameter to `createApp`, e.g. instead of

https://github.com/backstage/backstage/blob/d2af76ffb6a105dfec1310733f5ca8e316872116/packages/app/src/App.tsx#L116-L117

in your case you could have something like

```typescript
const app = createApp({
  apis: [
    createApiFactory({
      api: kubernetesAuthProvidersApiRef,
      deps: {
        cognitoAuthApi: cognitoAuthApiRef,
        googleAuthApi: googleAuthApiRef,
        microsoftAuthApi: microsoftAuthApiRef,
      },
      factory: ({
        cognitoAuthApi,
        googleAuthApi,
        microsoftAuthApi,
      }) => {
        return new KubernetesAuthProviders({
          microsoftAuthApi,
          googleAuthApi,
          oidcProviders: {
            cognito: cognitoAuthApi,
          },
        });
      },
    }),
    ...apis,
  ],
```

Mind you, this exact snippet would change things so that _only_ `cognito` was a valid `oidcTokenProvider` but hopefully you get the idea.

There are no docs describing this, though -- do you mind if I label this as a docs issue, or do you think there's still something lacking in the code?

**Comment by markussiebert** [comment]
Hi, we were able to implement this as you described it! Thank you a lot!

One thing is odd, our from scratch auth provider does not show up under settings -> authentication providers and we have to login again every time we visit the kubernetes tab in our component. Do you have an idea, how we can decrease the logins for this provider?

Apart from this - yes - it's only docs!

**Comment by jamieklassen** [comment]
I think the set of authentication providers in the settings page is something that is controlled separately by the [user-settings plugin](https://github.com/backstage/backstage/blob/master/plugins/user-settings/README.md#props). However it sounds like your custom auth provider might not be retrieving/handling refresh tokens properly -- natural things to check are whether you're requesting the `offline_access` scope and whether an `oidc-refresh-token` cookie is present.

**Comment by markussiebert** [comment]
So, just checked - I have an cognito-refresh-token (the name of my auth provider is cognito).
I can't request 'offline' or 'offline_access' as a scope.
But as I have a refresh token in a cookie, it should work - but it does not. Any further ideas what I can check?

**Comment by jamieklassen** [comment]
Maybe this depends a bit on the frontend implementation you have for this auth provider - Did you follow [these docs](https://backstage.io/docs/auth/oidc/) and use [`OAuth2.create`](https://backstage.io/docs/reference/core-app-api.oauth2/)?

**Comment by markussiebert** [comment]
Yes, that's what we did:

```
createApiFactory({
    api: cognitoOIDCAuthApiRef,
    deps: {
      discoveryApi: discoveryApiRef,
      oauthRequestApi: oauthRequestApiRef,
      configApi: configApiRef,
    },
    factory: ({ discoveryApi, oauthRequestApi, configApi }) =>
      OAuth2.create({
        discoveryApi,
        oauthRequestApi,
        configApi,
        provider: {
          id: 'cognito',
          title: 'Cognito',
          icon: () => null,
        },
        environment: configApi.getOptionalString('auth.environment'),

        defaultScopes: ['openid', 'email', 'profile'],
        popupOptions: {
          // optional, used to customize login in popup size
          size: {
            fullscreen: true,
          },
          /**
           * or specify popup width and height
           * size: {
              width: 1000,
              height: 1000,
            }
           */
        },
      }),
  }),
```

**Comment by jamieklassen** [comment]
tl;dr: maybe you can check the expiry timestamp for the tokens you're getting from cognito, or make a closer inspection of the request your browser is (hopefully) making to the refresh endpoint for your auth provider?

I'm a little stumped -- I'll include a bunch of information that I tend to use in situations like this and hopefully it'll jog something for you. What should happen when you open the Kubernetes tab is:
1. The `OidcKubernetesAuthProvider` calls `getIdToken` on your custom `OpenIdConnectApi` implementation:
    https://github.com/backstage/backstage/blob/e4781ab1e92cfd5cc7c43d1d13bdebbc9e7215ee/plugins/kubernetes-react/src/kubernetes-auth-provider/OidcKubernetesAuthProvider.ts#L48
1. Since you're using `OAuth2`, it delegates to its `sessionManager`:
    https://github.com/backstage/backstage/blob/e4781ab1e92cfd5cc7c43d1d13bdebbc9e7215ee/packages/core-app-api/src/apis/implementations/auth/oauth2/OAuth2.ts#L199-L202
1. That particular session manager does have some caveats around forcing a refresh -- so maybe worth following [these docs](https://backstage.io/docs/auth/troubleshooting#stepping-through-authentication-manually) to figure out what exactly the `providerInfo.expiresAt` value is:
    https://github.com/backstage/backstage/blob/e4781ab1e92cfd5cc7c43d1d13bdebbc9e7215ee/packages/core-app-api/src/apis/implementations/auth/oauth2/OAuth2.ts#L141-L157
1. In any case, the first thing that session manager will try to do is use a refresh token to get the desired scopes:
    https://github.com/backstage/backstage/blob/e4781ab1e92cfd5cc7c43d1d13bdebbc9e7215ee/packages/core-app-api/src/lib/AuthSessionManager/RefreshingAuthSessionManager.ts#L100-L102
1. And in this case, that works by hitting `/api/auth/cognito/refresh` -- can you see that request in your browser's dev tools? Is it giving a 401?:    https://github.com/backstage/backstage/blob/e4781ab1e92cfd5cc7c43d1d13bdebbc9e7215ee/packages/core-app-api/src/lib/AuthConnector/DefaultAuthConnector.ts#L137-L150

FWIW we should probably distinguish this OIDC troubleshooting thread from the core topic of this issue, which is the lacking documentation for custom `KubernetesAuthProvidersApi` implementations... For future readers, from [this comment](https://github.com/backstage/backstage/issues/19836#issuecomment-1724903548) onwards we changed topic.

**Comment by jamieklassen** [comment]
Just saw https://github.com/backstage/backstage/issues/20322#issuecomment-1747271192 -- not sure if it's at all related, but it reminded me of this.

**Comment by markussiebert** [comment]
Thank you a lot for your Help!

I looked into the requests via networking tab and I think it looks fine:

GET to http://localhost:7007/api/auth/cognito/refresh?optional&scope=openid&env=development
with header containing the cookie with cognito-refresh-token

Response is 200 with

```json
{
    "backstageIdentity": {
        "token": "valid token from backstage",
        "expiresInSeconds": 3599,
        "identity": {
            "type": "user",
            "userEntityRef": "user:default/88a953a1-f1b0-4cb5-9589-ee991143cb83",
            "ownershipEntityRefs": [
                "user:default/88a953a1-f1b0-4cb5-9589-ee991143cb83"
            ]
        }
    },
    "providerInfo": {
        "idToken": "valid token with ...
             auth_time: 1696585562, => 10 minutes ago
             exp: 1696589507, => in 50 minutes
              iat 1696585908, => now"
        "accessToken": "valid token",
        "expiresInSeconds": 3600
    },
    "profile": {
        "displayName": "Markus Siebert",
        "email": "..."
    }
}
```
No scope in providerInfo ... 

But even with this successfull request on klicking on the kubernetes tab, the login window appears (after the refresh call was already successfull)

![image](https://github.com/backstage/backstage/assets/16692560/20582afc-1e9d-45c4-b144-1b88f663c8c5)


Just wanted to let you know ...  
Also tried the changes from the PR from your last comment - nothing changed

**Comment by markussiebert** [comment]
One additional information: today confirmed that the token from the refresh call response is valid. And it is - it's refreshed without clicking on the LOG IN Button, but somehow it isn't used.

We also see, that our refresh token is encrypted with:

  "cty": "JWT",
  "enc": "A256GCM",
  "alg": "RSA-OAEP"

don't know if this matters. 

We added some console.logs to the plugin auth node refresh call for OIDC, but this is only called for our gitlab integration - not for the cognito oauth.

**Comment by dariozachow** [comment]
After further investigation we found out, that the scope was missing, when calling the refresh method located in the OidcAuthProvider: https://github.com/backstage/backstage/blob/856d08057a96271abd608c19a16f21622b66fad4/plugins/auth-backend/src/providers/oidc/provider.ts#L127-L132 
Therefore we've added the following line: 
`tokenset.scope = req?.scope`
after calling the `client.refresh()` method, to manually add the missing scopes which are configured in our app-config.yaml.

With these changes applied no relogin is required after refreshing the page or the scheduled refresh is triggered.
Is this an Issue on our side or a bug in the auth-backend plugin?

**Comment by jamieklassen** [comment]
@dariozachow do you and @markussiebert work together/use the same setup? Or are you encountering the same symptom with a different configuration?

**Comment by dariozachow** [comment]
Yeah,  we are working with the same code and configuration of our backstage project.
Therefore having the same problems with the oidc login.
